modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+package engine
+
+import scala.concurrent.ExecutionContext
+
+import cats.implicits._
+import cats.effect.Effect
+
+sealed trait Step
+object Step {
+
+  // Whatever is shared among all Steps. Notice it doesn't // extend Step
+  // because it's meant to be used only internally
+  case class Core(
+    id: String,
+    config: Core.Configuration,
+    breakpoint: Boolean
+  )
+
+  object Core {
+    case class Configuration()
+  }
+
+  sealed trait F2 extends Step
+  object F2 {
+
+    case class Done(core: Core, fileId: String) extends F2
+
+    sealed trait Pending extends F2
+    object Pending {
+      case class Standard(core: Core) extends Pending
+      case class Flat(core: Core) extends Pending
+      case class Dark(core: Core) extends Pending
+    }
```

Comment:
These case classes represent the different types of Steps with different execution definitions. They are only here as an example, but they should be the same for Pending, Ongoing, etc...

BTW, it's not still clear how to better represent this deeply nested ADT. Perhaps `Step.F2.Flat.Pending` instead of `Step.F2.Pending.Flat`. The reason for the first option is that in a Sequence the pending Steps can be of any execution type like `Dark`, `Flat` but need to be always Pending.